ninja-core/src/main/java/ninja/cache/CacheProvider.java

```diff
@@ -25,84 +27,51 @@
- * 
- * 
- * @author ra
- *
  */
 @Singleton
 public class CacheProvider implements Provider<Cache> {
-
-    private final NinjaProperties ninjaProperties;
-    private final Injector injector;
-
-    private final Cache cache;
-    private final Logger logger;
-
+    static private final Logger logger = LoggerFactory.getLogger(CacheProvider.class);
+    
+    private final ImplFromPropertiesFactory<Cache> factory;
+    private final Supplier<Cache> supplier;
+    
     @Inject
-    public CacheProvider(
-                         Injector injector, 
-                         NinjaProperties ninjaProperties, 
-                         Logger logger) {
-        
-        this.ninjaProperties = ninjaProperties;
-        this.injector = injector;
-        this.logger = logger;
+    public CacheProvider(Injector injector, NinjaProperties ninjaProperties) {
+        this.factory = new ImplFromPropertiesFactory<>(
+            injector,
+            ninjaProperties,
+            NinjaConstant.CACHE_IMPLEMENTATION,
+            Cache.class,
+            "ninja.cache.CacheEhCacheImpl",
+            true,
+            logger);
         
-      
-        Class<? extends Cache> cacheClass = null;
-
-        String cacheImplClassName = ninjaProperties.get(NinjaConstant.CACHE_IMPLEMENTATION);
-        
-        if (cacheImplClassName != null) {
-            try {
-
-                Class<?> clazz = Class.forName(cacheImplClassName);
-                cacheClass = clazz.asSubclass(Cache.class);
-
-                logger.info("Using the {} as implementation for caching.",  cacheClass);
-
-            } catch (ClassNotFoundException e) {
-                
-                throw new RuntimeException(
-                        "Class defined in configuration " + NinjaConstant.CACHE_IMPLEMENTATION +
-                        "not found (" + cacheClass + ")", e);
-                
-            } catch (ClassCastException e) {
-                
-                throw new RuntimeException(
-                        "Class defined in configuration " 
-                                + NinjaConstant.CACHE_IMPLEMENTATION +
-                                "is not an instance of interface cache (" 
-                                + cacheClass + ")", e);
+        // lazy singleton
+        this.supplier = Suppliers.memoize(new Supplier<Cache>() {
+            @Override
+            public Cache get() {
+                return factory.create();
             }
-        }
-
-        if (cacheClass == null) {
-                // load default implementation
-                cacheClass = CacheEhCacheImpl.class;
-                logger.info("Using default eh cache implementation. ({}) ", cacheClass);
-            
-        }
-
-        cache = injector.getInstance(cacheClass);
-      
+        });
     }
 
     @Override
     public Cache get() {
-        // only called once => reference cached by guice...
-        return cache;
-        
+        return this.supplier.get();
```

Comment:
I had thought so as well.  The problem is that Cache, Postoffice, etc. should be Singletons.  Guice does not allow Singletons to be lazily built on-demand.  So let's say you don't use Cache anywhere in your app -- Guice would still build it & start it due to be being bound as a singleton.  Postoffice is worse since it validates you have a config key of 'smpt.host' set, but even if you don't use it in your app, your app won't start.  So the solution was to not bind them as Singleton, and then lazily create a singleton instance with a supplier.  If you then inject a Cache or Postoffice into your controller, it'll fail on start, but if you never do then they are essentially ignored and never built.